File: VB/ExpandoObject/ViewModels/MainViewModel.vb
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports DevExpress.Mvvm.CodeGenerators

Namespace ExpandoObject.ViewModels

    <GenerateViewModel>
    Public Partial Class MainViewModel

        <GenerateProperty>
        Private _items As ObservableCollection(Of Dynamic.ExpandoObject)

        Public Sub New()
            _items = New ObservableCollection(Of Dynamic.ExpandoObject)()
            For i = 0 To 10 - 1
                Dim item As dynamic = New Dynamic.ExpandoObject()
                item.Id = i
                item.Name = $"Item {i}"
                item.CreatedAt = Date.Now.AddDays(i)
                _items.Add(item)
            Next
        End Sub

        <GenerateCommand>
        Public Sub AddColumn()
            For Each item In _items
                Dim dict = CType(item, IDictionary(Of String, Object))
                dict.Add($"Value {dict.Keys.Count}", _items.IndexOf(item) * dict.Keys.Count)
            Next
        End Sub
    End Class
End Namespace

File: CS/ExpandoObject/ViewModels/MainViewModel.cs
using DevExpress.Mvvm;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Windows.Input;

namespace ExpandoObject.ViewModels {
    public class MainViewModel : ViewModelBase {
        public ObservableCollection<System.Dynamic.ExpandoObject> Items { get => GetProperty(() => Items); set => SetProperty(() => Items, value); }

        public MainViewModel() {
            Items = new ObservableCollection<System.Dynamic.ExpandoObject>();

            for (var i = 0; i < 10; i++) {
                dynamic item = new System.Dynamic.ExpandoObject();
                item.Id = i;
                item.Name = $"Item {i}";
                item.CreatedAt = DateTime.Now.AddDays(i);

                Items.Add(item);
            }

            AddColumnCommand = new DelegateCommand(AddColumn);
        }

        public void AddColumn() {
            foreach (var item in Items) {
                var dict = (IDictionary<string, object>)item;
                dict.Add($"Value {dict.Keys.Count}", Items.IndexOf(item) * dict.Keys.Count);
            }
        }

        public ICommand AddColumnCommand { get; }
    }
}